### PR TITLE
Task termination

### PR DESCRIPTION
**Pull request by satoshikumano**
Enable to terminate task.

Add `continue callback` determines whether to continue or terminate. Continue callback is called when memory/ resources managed by SDK is freed. If termination is indicated by the return value of the callback, `exit callback` is called and application can free memory/ resources managed by application and execute necessary steps exiting task (eg. call `vTaskDelete ()` if using Free RTOS)

## New callbacks

`void * task_info` actual type is determined by SDK and varies depending on tasks.
Application will cast to the type to obtain task specific information if needed.

```c
typedef void
(*KII_TASK_EXIT)
    (void* task_info,
    void* userdata);

typedef kii_bool_t
(*KII_TASK_CONTINUE)
    (void* task_info,
    void* userdata);
```
## New `kii` APIS
```c
void kii_set_task_continue_cb(kii_t* kii, KII_TASK_CONTINUE continue_cb, void* userdata);
void kii_set_task_exit_cb(kii_t* kii, KII_TASK_EXIT exit_cb, void* userdata);
```

## New `tio` APIs
```c
void tio_handler_set_cb_task_continue(tio_handler_t* handler, KII_TASK_CONTINUE cb_continue, void* userdata);
void tio_handler_set_cb_task_exit(tio_handler_t* handler, KII_TASK_EXIT cb_exit, void* userdata);
void tio_updater_set_cb_task_continue(tio_updater_t* updater, KII_TASK_CONTINUE cb_continue, void* userdata);
void tio_updater_set_cb_task_exit(tio_updater_t* updater, KII_TASK_EXIT cb_exit, void* userdata);
```

## Notes about `tio_updater`

Change `tio_upadater` task loop to call continue callback every one second in 1d31b6b
.
If continue callback is called only when `state upload interval` elapsed, application needs to wait for long time to terminate `tio_upadater` task.

### CODE REVIEW

File: tio/include/tio.h [hunk 1]
```diff
@@ -177,6 +190,64 @@ void tio_handler_set_mqtt_to_sock_recv(tio_handler_t* handler, unsigned int to_s
...
+ * In addition, you may need to call task/ thread termination API.
+ * It depends on the task/ threading framework you used to create task/ thread.
+ * After the exit_cb returned, task function immediately returns.
+
+ * If this API is not called or set NULL,
+ * task function immediately returns when task is discontinued or un-recoverble error occurs.
+
+ * \param handler [out] tio_handler_t instance.
+ * \param cb_exit [in] Called right before the exit. task_info argument type is tio_handler_task_info*
+ * \param userdata [in] Context data pointer passed as second argument when cb_exit is called.
```
Comment (naoki-akimoto) [inline tio/include/tio.h]: `tio_handler_set_cb_task_continue` の方だと `when KII_TASK_CONTINUE callback is called.` となってるけどこっちでは `when cb_exit is called.` となっていて統一感が無いのでどちらかに合わせましょう。

File: kii/kii.h [hunk 2]
```diff
@@ -638,6 +637,57 @@ int kii_set_mqtt_cb_sock_close(kii_t* kii, KHC_CB_SOCK_CLOSE cb, void* userdata)
...
+ * - kii_set_task_continue_cb()
+ * - kii_set_task_exit_cb()
+
+ * In addition, you may need to call task/ thread termination API.
+ * It depends on the task/ threading framework you used to create task/ thread.
+ * After the exit_cb returned, task function immediately returns.
+
+ * If this API is not called or set NULL,
+ * task function immediately returns when task is discontinued or un-recoverble error occurs.
+ */
```
Comment (naoki-akimoto) [inline kii/kii.h]: `param` の記載がありません。

File: tio/include/tio.h (hunk 1 above)
Comment (naoki-akimoto) [reply tio/include/tio.h]: updater側だと `when KII_TASK_* callback is called.` なのでこちらに統一しましょう。

Comment (satoshikumano) [reply tio/include/tio.h]: `KII_TASK_* callback` は型なので、実際にコールされるメソッドのインスタンスを指すように修正しました。

**Pull request merged**